### PR TITLE
Fix DCE on mutable calls in a loop.

### PR DESCRIPTION
**Pull request by csyonghe**
This change fixes a DCE bug that mistakenly removes a call to a mutable function inside a loop.
A test case that reproduces the bug is included and fixed with this change.

### CODE REVIEW

File: source/slang/slang-ir-propagate-func-properties.cpp [hunk 1]
```diff
@@ -238,34 +241,37 @@ class NoSideEffectFuncPropertyPropagationContext : public FuncPropertyPropagatio
...
+                case kIROp_unconditionalBranch:
+                case kIROp_Switch:
+                case kIROp_Return:
+                case kIROp_loop:
+                case kIROp_Call:
+                case kIROp_Param:
+                case kIROp_Unreachable:
+                case kIROp_Store:
+                case kIROp_SwizzledStore:
+                    break;
```
Comment (tangent-vector) [inline source/slang/slang-ir-propagate-func-properties.cpp:256]: The duplication here with the logic above seems unfortunate. Is there any way to factor this out into a subroutine?

File: source/slang/slang-ir-util.cpp [hunk 2]
```diff
@@ -688,9 +693,45 @@ bool areCallArgumentsSideEffectFree(IRCall* call)
...
+                        {
+                            callBlock = dom->getImmediateDominator(callBlock);
+                            if (!callBlock)
+                                return false; // If we are here, var does not dominate the call, which should never happen.
+                            if (as<IRLoop>(callBlock->getTerminator()))
+                                return false; // The var is used in a loop, must return false.
+                        }
+                        // If we reach here, the var is used as an inout parameter for the call, but the call
+                        // is not nested in a loop at an higher nesting level than where the var is defined,
+                        // so we can treat the use as side effect free.
```
Comment (tangent-vector) [inline source/slang/slang-ir-util.cpp:736]: I like this analysis, although it starts to feel like our definition of "side effect free" is getting a bit loose. What we really mean here is DCE-able and/or CSE-able instructions, right?

File: source/slang/slang-ir.cpp [hunk 3]
```diff
@@ -7082,6 +7098,8 @@ namespace Slang
                 module->getDeduplicationContext()->getConstantMap().remove(IRConstantKey{ constInst });
             }
             module->getDeduplicationContext()->getInstReplacementMap().remove(this);
+            if (auto func = as<IRGlobalValueWithCode>(this))
+                module->invalidateDominatorTree(func);
```
Comment (tangent-vector) [inline source/slang/slang-ir.cpp:7102]: Could this be a more general `module->invalidateAnalysis(this)`?

File: source/slang/slang-ir.h [hunk 4]
```diff
@@ -2072,6 +2080,17 @@ struct IRModule : RefObject
...
 
+    IRDominatorTree* findDominatorTree(IRGlobalValueWithCode* func)
+    {
+        IRAnalysis* analysis = m_mapInstToAnalysis.tryGetValue(func);
+        if (analysis)
+            return analysis->getDominatorTree();
+        return nullptr;
+    }
+    IRDominatorTree* findOrCreateDominatorTree(IRGlobalValueWithCode* func);
+    void invalidateDominatorTree(IRGlobalValueWithCode* func) { m_mapInstToAnalysis.remove(func); }
```
Comment (tangent-vector) [inline source/slang/slang-ir.h:2091]: Yeah, it seems like this one could be `invalidateAnalysisForInst(...)`.

**Review by tangent-vector — DISMISSED** [review]

File: source/slang/slang-ir-propagate-func-properties.cpp (hunk 1 above)
Comment (csyonghe) [reply source/slang/slang-ir-propagate-func-properties.cpp]: Done.

File: source/slang/slang-ir-util.cpp (hunk 2 above)
Comment (csyonghe) [reply source/slang/slang-ir-util.cpp]: Done.

File: source/slang/slang-ir.cpp (hunk 3 above)
Comment (csyonghe) [reply source/slang/slang-ir.cpp]: Done.

File: source/slang/slang-ir.h (hunk 4 above)
Comment (csyonghe) [reply source/slang/slang-ir.h]: Done.

**Pull request merged**